File: src/javascript/algorithms/dynamicProgramming/longestIncreasingSequence.js
// TC:O(n * 2) SC:O(n)

function lengthOfLIS(nums) {
    let dp = new Array(nums.length).fill(0);
    dp[0] = 1;
    let maxLength = 0;
    for(let i=1; i< nums.length; i++) {
        let len = 0;
        for(let j=0; j<i; j++){
            if(nums[j]< nums[i]) {
                len = Math.max(len, dp[j]);
            }
            dp[i] = 1 + len;
        }
        maxLength = Math.max(maxLength, dp[i]);
    }
    return maxLength;
}

let nums = [10,9,2,5,3,7,101,18];
console.log(lengthOfLIS(nums));

File: src/java1/algorithms/dynamicProgramming/longestIncreasingSequence/LongestIncreasingSequence.java
// TC:O(n * 2) SC:O(n)
package java1.algorithms.dynamicProgramming.longestIncreasingSequence;

import java.util.Arrays;

public class LongestIncreasingSequence {
    private static int lengthOfLIS(int[] nums) {
        int[] dp = new int[nums.length];
        Arrays.fill(dp, 1);
        int maxLength = 0;
        for(int i = nums.length-1; i>=0; i--) {
            for(int j=i+1; j<nums.length; j++) {
                if(nums[i] < nums[j]) {
                    dp[i] = Math.max(dp[i], 1+dp[j]);
                }
            }
            maxLength = Math.max(maxLength, dp[i]);
        }
        return maxLength;
    }

    public static void main(String[] args) {
        int[] nums1 = {9,7,1,4,2,6,10,12};
        int[] nums2 = {1,3,4,6,7,8};
        int[] nums3 = {5,4,3,2,1};
        System.out.println(lengthOfLIS(nums1));
        System.out.println(lengthOfLIS(nums2));
        System.out.println(lengthOfLIS(nums3));
    }
}
